### ISSUE TITLE
SwiftPM build tool source target file list is stale

### ISSUE THREAD

**Issue by marcprux** [apple/swift-package-manager#6816]
### Description

PackagePlugin.Target's sourceFiles list is incorrectly cached and passed to `createBuildCommands`, and doesn't accurately reflect external changes to the project's build graph until Xcode is quit or restarted. This makes SwiftPM build plugins unsuitable for pure Swift projects that add, remove, or rename files in their projects without frequent restarting of Xcode.

### Expected behavior

The up-to-date file list should be sent to the build plugin.

### Actual behavior

The list of files passed to `BuildToolPlugin.createBuildCommands` does not reflect any changes to the file system since the plugin was first run. Files that have been delete are still included in the `PackagePlugin.SourceModuleTarget.sourceFiles` list, and new files are not added. This results in errors for any plugin that relies on the list of files being an accurate representation of the file system.

### Steps to reproduce

  1. Clone https://github.com/marcprux/SPMFileCacheBug.git
  2. Open the Package.swift in Xcode (`xed Package.swift`)
  3. Build the "SPMFileCacheBug" target successfully
  4. Select the Sources/SPMFileCacheBug/SPMFileCacheBug.swift file and run the File->Duplicate… menu
  5. Build the "SPMFileCacheBug" target, observing the plugin fail with the error:

```error
swiftSourceFiles: ["SPMFileCacheBug.swift"]
actualSwiftSourceFiles: ["SPMFileCacheBug.swift", "SPMFileCacheBug copy.swift"]
PackagePlugin.Target returned wrong results
```

### Swift Package Manager version/commit hash

5.9

### Swift & OS version (output of `swift --version ; uname -a`)

```
swift-driver version: 1.87.1 Apple Swift version 5.9 (swiftlang-5.9.0.128.2 clang-1500.0.40.1)
Target: arm64-apple-macosx13.0
Darwin zap.local 22.6.0 Darwin Kernel Version 22.6.0: Wed Jul  5 22:22:05 PDT 2023; root:xnu-8796.141.3~6/RELEASE_ARM64_T6000 arm64
```

**Comment by neonichu** [comment]
What happens if you manually reload the package graph via the menu item in Xcode? Does that help or are you still getting the cached values?

**Comment by neonichu** [comment]
I saw FB12969712 and it looks like this is indeed an Xcode specific issue which can be worked around with manual reloading. So I'm going to close this since this cannot be addressed in SwiftPM.

**Comment by abramchuk-v** [comment]
@neonichu 
FB12969712 is some internal ID for this issue?

**Comment by MaxDesiatov** [comment]
FB12969712 is an ID of the issue filed in Feedback Assistant.

**Comment by abramchuk-v** [comment]
I got, I mean we have no option to check out issue with this number, how can we keep track or raise up this issue?

**Comment by tomerd** [comment]
cc @abertelrud

**Comment by neonichu** [comment]
To clarify, I related these issues because it seemed like both were filed by @marcprux, but either way, since this is an Xcode specific issue, it doesn't really make sense to track it here, we cannot fix it.

You can file your own report through https://feedbackassistant.apple.com to raise awareness of the issue.

**Comment by abramchuk-v** [comment]
Thanks for clarification!

**Comment by marcprux** [comment]
For the record, this has been fixed in Xcode 16.0.
